pkg/docker-exec-ws/shellclient.go

```diff
@@ -0,0 +1,355 @@
+package dockerExecWS
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/taskcluster-cli/pkg/iopipes"
+	"github.com/taskcluster/taskcluster-worker/plugins/interactive/shellconsts"
+	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
+)
+
+const (
+	// MaxOutstandingBytes is the maximum number of unsent bytes in the stdin buffer.
+	MaxOutstandingBytes = 8 * 1024 * 1024
+	pingInterval        = 15 * time.Second
+	writeTimeout        = 2 * pingInterval
+	readTimeout         = 3 * pingInterval
+)
+
+var dialer = websocket.Dialer{
+	HandshakeTimeout: shellconsts.ShellHandshakeTimeout,
+	ReadBufferSize:   shellconsts.ShellMaxMessageSize,
+	WriteBufferSize:  shellconsts.ShellMaxMessageSize,
+}
+
+// ShellClient is the implementation of
+// github.com/taskcluster/taskcluster-worker/engines.Shell for v1 shells.
+type ShellClient struct {
+	ws       *websocket.Conn
+	done     chan struct{}
+	mWrite   sync.Mutex
+	cPause   *sync.Cond
+	paused   bool
+	complete atomics.Once
+	success  bool
+	err      error
+
+	stdin        io.WriteCloser
+	stdinReader  io.ReadCloser
+	stdout       io.ReadCloser
+	stdoutWriter *iopipes.DrainingPipeWriter
+	stderr       io.ReadCloser
+	stderrWriter *iopipes.DrainingPipeWriter
+}
+
+// Dial will built a proper shell websocket URL, connect to the server and return an initialized shell client.
+func Dial(socketURL string, command []string, tty bool) (*ShellClient, error) {
+	u, err := url.Parse(socketURL)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid socketURL: %s, parsing error: %s",
+			socketURL, err)
+	}
+	q := u.Query()
+
+	// Ensure the URL has ws or wss as scheme
+	switch u.Scheme {
+	case "http":
+		u.Scheme = "ws"
+	case "https":
+		u.Scheme = "wss"
+	} // Set command arguments overwriting any existing querystring values
+	q.Del("command")
+	for _, arg := range command {
+		q.Add("command", arg)
+	}
+
+	// Set tty true or false
+	if tty {
+		q.Set("tty", "true")
+	} else {
+		q.Set("tty", "false")
+	}
+
+	// Set querystring on url
+	u.RawQuery = q.Encode()
+
+	// Dial up to the constructed URL
+	ws, _, err := dialer.Dial(u.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return a new ShellClient wrapping the websocket
+	return New(ws), nil
+}
+
+// New creates a new v1 ShellClient from an existing websocket.
+func New(ws *websocket.Conn) *ShellClient {
```

Comment:
Should we expose this? when `Dial` is the interface people should use?

Reply:
I left it exposed because it is to match the v2 client (https://godoc.org/github.com/taskcluster/taskcluster-worker/plugins/interactive/shellclient) that has it exposed.

Reply:
oh, yeah, I guess it might be useful from time to time...

I think I needed in the worker, because I do some really crazy things with websockets :)